### PR TITLE
Add BearerTokenPrincipal

### PR DESCRIPTION
**Pull request by cbanek**
Add a BearerTokenPrincipal, which is just the authorization header
set with something that starts with Bearer.  This is automatically
pulled out in the ServletPrincipalExtractor and presented as a
Principal on the subject.

### CODE REVIEW

**Comment by cbanek** [comment]
Excellent - totally agree on tests, I just wanted to check myself before I get too far down the rabbit hole.  Version number is also not a problem at all!

**Comment by pdowler** [comment]
If you look at the cookie handling you see that it conveys a username and after parsing it creates an HttpPrincipal (yeah - poorly named because we started with HTTP Auth which is username/password). I was expecting oauth to be somewhat like this... 

What is the identity that  is conveyed in a bearer token?

**Comment by cbanek** [comment]
Sadly the bearer token is pretty much opaque.  It just means if you have it, that's all you need.  So there's not really a great way to know what kind of format it is, although in my case, it's a JWT token, which I parse out in my AuthorizerImpl over here:  https://github.com/lsst-sqre/lsst-tap-service/pull/27

This is where I'm taking the BearerTokenPrincipal and changing it over to a set of principals with the info in them.  But there isn't a way to say when you have a token in the auth header what type it is as far as I know.  So that's why I pretty much just stashed the token, and am doing the parsing somewhere else.

If I could parse out the token in the PrincipalExtractor, I think that would make more sense, but that would probably put a little too much of my LSST business logic in that mostly generic layer of code.  Even if we knew it was a JWT token, the claims and fields in it are just a JSON dict and are not necessarily standardized over all JWT tokens, other than that they are also JSON dicts.

I think another way of doing this might be to extend the PrincipalExtractor to allow for me to add maybe some delegates, like, a function to add a new object to call an extract function on.  This would allow me to insert myself into the ServletPrincipalExtractor, which seems to not be as extensible as the AuthorizerImpl (although we could do that).

**Comment by pdowler** [comment]
So bearer token is an opaque object that can be turned into identity(ies) by calling back to a service. In that case I think you are on the right track by just capturing it in the PrincipalExtractor.

The next thing would be to implement your own AuthenticatorImpl which takes the Subject with the token in it and "augment" it with real valid identity(ies). This is where you would call back to your auth-support service and is the intended plugin. Note that while our impl just adds more principals, yours could (as necessary): 
* create and return a new Subject instance
* remove the BearerTokenPrincipal if it is no longer useful as a temp identity
* store the bearer token in the credentials if it could be re-used for other calls (might need some sort of wrapper code like cookies to limit scope where it can/should be used)

File: cadc-util/src/main/java/ca/nrc/cadc/auth/BearerTokenPrincipal.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+     *
+     * @return the name of this principal.
+     */
+    @Override
+    public String getName()
+    {
+        return getToken();
+    }
+
+    public String getToken()
```
Comment (pdowler) [inline cadc-util/src/main/java/ca/nrc/cadc/auth/BearerTokenPrincipal.java]: I probably would not bother with getToken() and just make the javadoc for getName() clear that it returns the entire token.

File: cadc-util/src/main/java/ca/nrc/cadc/auth/ServletPrincipalExtractor.java [hunk 2]
```diff
@@ -164,6 +164,12 @@ public ServletPrincipalExtractor(final HttpServletRequest req)
             finally { }
         }
 
+        String authToken = request.getHeader("Authorization");
```
Comment (pdowler) [inline cadc-util/src/main/java/ca/nrc/cadc/auth/ServletPrincipalExtractor.java]: I would declare a constant like AuthenticationUtil.HTTP_AUTHORIZATION_HEADER with this value and document it as being defined in HTTP/1.1

**Comment by cbanek** [comment]
Yep I think I've done that in the PR on my side in lsst-sqre/lsst-tap-service#27 although I don't have to call back to a service because everything is encoded in the token.  But I am making a NumericPrincipal (with the UID, although it's a bit strange that it's a UUID, but I think you're just using it as a packing method), and an HttpPrincipal with the username for it.  I'm also keeping around the original BearerTokenPrincipal because if anything needs to call out, then I could use that token on those calls to do impersonation (as long as the tokens are valid).

File: cadc-util/src/main/java/ca/nrc/cadc/auth/BearerTokenPrincipal.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
+/*
```
Comment (pdowler) [inline cadc-util/src/main/java/ca/nrc/cadc/auth/BearerTokenPrincipal.java]: This isn't the full header comment that includes the license terms. You can find the correct one in AuthenticationUtil.

**Comment by pdowler** [comment]
In that case, I would remove the BearerTokenPrincipal from the principals set and store the token in the credentials set for later re-use. Cookies (tokens) are wrapped and include a domain where they chould be sent; we also need to do that with certificate chains and I assume bearer tokens as well. In the past we assumed a single cert was usable everywhere (IVOA SSO-1.x) but that dream seems to have died. BTW, proxy certs get created/stored in the subject's credential set elsewhere (cadc-cdp, CredUtil) but the cookie example should be there.

**Comment by cbanek** [comment]
Okay, pushed a new commit with changes from feedback, added some simple tests, and bumped the version number!  Ready for round 2.

**Comment by pdowler** [comment]
published cadc-util-1.2.18 to jcenter maven repository

**Pull request merged**